flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java

```diff
@@ -56,7 +56,7 @@
 public class ChannelStateWriterImpl implements ChannelStateWriter {
 
 	private static final Logger LOG = LoggerFactory.getLogger(ChannelStateWriterImpl.class);
-	private static final int DEFAULT_MAX_CHECKPOINTS = 5; // currently, only single in-flight checkpoint is supported
+	private static final int DEFAULT_MAX_CHECKPOINTS = 1000; // includes max-concurrent-checkpoints + checkpoints to be aborted (scheduled via mailbox)
```

Comment:
I guess the current way seems a temporary work-around solution, not an elegant way. The initial purpose for introducing this threshold is for logic validating and avoiding invalid checkpoints retained in writer forever. But if we consider the abort delay into this threshold, it seems somehow lose the initial meaning for the guard, and we are really not sure what is the proper value for this threshold.

The proper way might resolve the potential race condition in essence, but it might pay more efforts not feasible ATM. So I think we might leave another debt here in future.

Reply:
Me and @AHeise also had these concerns and discussed it [above](https://github.com/apache/flink/pull/12478#discussion_r436502262) (I increased the limit to 1K since then).

I'm not sure if there __is__ a proper solution because in the general case (max-concurrent-checkpoints > 1) we simply don't know which checkpoints to abort. Another issue is thread-safety, which is probably solvable but with some extra complexity which I think doesn't pay off here.

Reply:
Ok, let's keep it as now since no better solutions right now. And further improve it if possible future.